From 8ce380ebf60fbefa031179e71466acfd8f8639b3 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 28 Nov 2018 17:04:59 +0800
Subject: [PATCH 06/18] qwaylanddisplay: Refactor flushRequests() for multiple
 readers

Cancel reading when no events polled, since the other clients, e.g.
waylandsink, might already read the events before us.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 src/client/qwaylanddisplay.cpp | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/src/client/qwaylanddisplay.cpp b/src/client/qwaylanddisplay.cpp
index aef31961..69c8fff6 100644
--- a/src/client/qwaylanddisplay.cpp
+++ b/src/client/qwaylanddisplay.cpp
@@ -74,6 +74,7 @@
 #include <QtCore/QDebug>
 
 #include <errno.h>
+#include <poll.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -187,11 +188,23 @@ void QWaylandDisplay::checkError() const
 
 void QWaylandDisplay::flushRequests()
 {
-    if (wl_display_prepare_read(mDisplay) == 0) {
+    struct pollfd fd;
+    int ret;
+
+    while (wl_display_prepare_read(mDisplay) != 0)
+        wl_display_dispatch_pending(mDisplay);
+
+    fd.fd = wl_display_get_fd(mDisplay);
+    fd.events = POLLIN | POLLERR | POLLHUP;
+    ret = poll(&fd, 1, 0);
+    if (ret > 0 && fd.revents & POLLIN) {
         wl_display_read_events(mDisplay);
+        wl_display_dispatch_pending(mDisplay);
+    } else {
+        wl_display_cancel_read(mDisplay);
     }
 
-    if (wl_display_dispatch_pending(mDisplay) < 0) {
+    if (wl_display_get_error(mDisplay) != 0) {
         checkError();
         exitWithError();
     }
-- 
2.20.1

